tasks/DroneFlightControlTask.cpp

```diff
@@ -0,0 +1,440 @@
+    mMaxVelocity = 10;
+    mIdleVelocity = 5;
+    mFinishAction = 0;
+    mExecutiveTimes = 1;
+    mYawMode = 0;
+    mTraceMode = 0;
+    mRCLostAction = 1;
+    mGimbalPitch = 0;
+    mLatitude = 0;
+    mLongitude = 0;
+    mAltitude = 0;
+
+    if (!initVehicle())
+        return false;
+    if (!checkTelemetrySubscription())
+        return false;
+    if (!missionInitSettings())
+        return false;
+
+    return true;
+}
+
+bool DroneFlightControlTask::startHook()
+{
+    if (!DroneFlightControlTaskBase::startHook())
+        return false;
+
+    // Obtain Control Authority
+    mVehicle->obtainCtrlAuthority(mFunctionTimeout);
+    return true;
+}
+
+typedef DroneFlightControlTask::States TaskState;
+static TaskState djiStatusFlightToTaskState(uint8_t status)
+{
+    switch (status)
+    {
+    case DJI::OSDK::VehicleStatus::FlightStatus::STOPED:
+        return TaskState::DJI_STOPPED;
+    case DJI::OSDK::VehicleStatus::FlightStatus::ON_GROUND:
+        return TaskState::DJI_ON_GROUND;
+    case DJI::OSDK::VehicleStatus::FlightStatus::IN_AIR:
+        return TaskState::DJI_IN_AIR;
+    }
+    // Never reached
+    throw std::invalid_argument("invalid controller state");
+}
+
+void DroneFlightControlTask::updateHook()
+{
+    _pose_samples.write(getRigidBodyState());
+    _battery.write(getBatteryStatus());
+
+    // cmd input
+    COMMAND_ACTION cmd_input;
+    if (_cmd_input.read(cmd_input) != RTT::NoData)
+        return;
+
+    // setpoint input
+    VehicleSetpoint cmd_pos;
+    if (_cmd_pos.read(cmd_pos) != RTT::NewData)
+        return;
+
+    // Check status
+    auto djiStatusFlight = mVehicle->subscribe->getValue<Telemetry::TOPIC_STATUS_FLIGHT>();
+    DroneFlightControlTask::States status = djiStatusFlightToTaskState(djiStatusFlight);
+    if (state() != status)
+        state(status);
+
+    switch (cmd_input)
+    {
+    case TAKEOFF_ACTIVATE:
+        return takeoff(cmd_pos);
+        if (djiStatusFlight == VehicleStatus::FlightStatus::IN_AIR)
+        {
+            case PRE_LANDING_ACTIVATE:
+                return preLand(cmd_pos);
+            case LANDING_ACTIVATE:
+                return land();
+            case GO_TO_ACTIVATE:
+                return goTo(cmd_pos, getRigidBodyState());
+            case MISSION_ACTIVATE:
+                return mission();
+        }
+    }
+
+    DroneFlightControlTaskBase::updateHook();
+}
+void DroneFlightControlTask::errorHook()
+{
+    DroneFlightControlTaskBase::errorHook();
+}
+void DroneFlightControlTask::stopHook()
+{
+    mVehicle->releaseCtrlAuthority(mFunctionTimeout);
+
+    DroneFlightControlTaskBase::stopHook();
+}
+void DroneFlightControlTask::cleanupHook()
+{
+    int pkgIndex = 0;
+    /*! Telemetry subscription*/
+    if (!teardownSubscription(pkgIndex))
+        DERROR("Failed to tear down Subscription!");
+
+    DroneFlightControlTaskBase::cleanupHook();
+}
+
+bool DroneFlightControlTask::initVehicle()
+{
+    bool threadSupport = true;
```

Comment:
Best way to manage the Vehicle pointer here would be to use a local unique_ptr, and convert mVehicle in a unique_ptr as well

~~~
{
   std::unique_ptr<Vehicle> vehicle(new Vehicle(...));
   ...
   // end of method
   mVehicle = std::move(vehicle)
}
~~~